Copy files

yarondbb opened the pull request.

yarondbb committed:
Solution is: Add a management layer above the copy queue:
The layer gets all file copy requests
The layer has waiting list of files to be copied. If more files are requested, they are merged into the waiting list.
The layer manages the files copy so that at any given time only 'max_copy_streams' are open.
The layer has a thread which moves files from waiting list to start copy if streams are availalbe

yarondbb committed:
Solution is: Add a management layer above the copy queue:
The layer gets all file copy requests
The layer has waiting list of files to be copied. If more files are requested, they are merged into the waiting list.
The layer manages the files copy so that at any given time only 'max_copy_streams' are open.
The layer has a thread which moves files from waiting list to start copy if streams are availalbe

yarondbb committed:
Solution is: Add a management layer above the copy queue:
The layer gets all file copy requests
The layer has waiting list of files to be copied. If more files are requested, they are merged into the waiting list.
The layer manages the files copy so that at any given time only 'max_copy_streams' are open.
The layer has a thread which moves files from waiting list to start copy if streams are availalbe

bbfsdev commented:
I see you handled all the things we talked about + moved the tcp call outside.
Good work, please merge the changes.

yarondbb committed:
Fixed issue where stream was not deleted correctly and caused a gem in the files under copy data base.

when file recieved on backup we used to check the cynamic content data. Now we check directly in the destination location. This will pass an issue where the monitoring on backup is slow and caused files already saved to destination be requested again.

yarondbb committed:
New design which uses a queue to start copy files per demend (when there are waiting files to be copied and when there is an open place in streams

yarondbb commented:
I have enhanced the design so it will use a queue for copy on demend (as you wanted).
This is probalby the fastest way to handle it.
Pls review the new code.
Also I have changed the way backup is checking if file already exists - instead of checking the dynamic content data it is now chaking the actuall destination of the file (so we are not dependent on cycles of monitoring which might be slow)

yarondbb commented:
@kolmanv  pls review again

yarondbb committed:
remove the use if sleep (in start copy) and moved to queue only
